test/unit/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamWriterTest.java

```diff
@@ -126,43 +152,98 @@ public void testBlockWriterOverWire() throws IOException
+     * redistribution task will modify index summary causing actual transfered index summary size to be different from
+     * the one in {@link ComponentManifest}
+     */
+    @Test
+    public void testBlockReadingAndWritingOverWireWithIndexSummaryRedistribution() throws Exception
+    {
+        testBlockReadingAndWritingOverWire(() -> {
+
+            long nonRedistributingOffHeapSize = 0;
+            long memoryPoolBytes = 1024 * 1024;
+
+            // rewrite index summary file with new min/max index interval
+            TableMetadata origin = store.metadata();
+            MigrationManager.announceTableUpdate(origin.unbuild().minIndexInterval(1).maxIndexInterval(2).build(), true);
+
+            try (LifecycleTransaction txn = store.getTracker().tryModify(asList(sstable), OperationType.UNKNOWN))
+            {
+                IndexSummaryManager.redistributeSummaries(new IndexSummaryRedistribution(ImmutableMap.of(store.metadata().id, txn),
+                                                                                         nonRedistributingOffHeapSize,
+                                                                                         memoryPoolBytes));
+            }
+            descriptor.getMetadataSerializer().mutateRepairMetadata(descriptor, 0, UUID.randomUUID(), false);
 
-        CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, CassandraOutgoingFile.getComponentManifest(sstable));
+            // reset min/max index interval
+            MigrationManager.announceTableUpdate(origin, true);
 
+            return null;
+        });
+    }
+
+    public void testBlockReadingAndWritingOverWire(Callable<?> componentMutation) throws Exception
+    {
         // This is needed as Netty releases the ByteBuffers as soon as the channel is flushed
         ByteBuf serializedFile = Unpooled.buffer(8192);
-        EmbeddedChannel channel = createMockNettyChannel(serializedFile);
-        AsyncStreamingOutputPlus out = new AsyncStreamingOutputPlus(channel);
+        InetAddressAndPort peer = FBUtilities.getBroadcastAddressAndPort();
+        StreamSession session = setupStreamingSessionForTest();
+        ComponentManifest manifest = CassandraOutgoingFile.getComponentManifest(sstable);
+        CassandraStreamHeader header = header(manifest);
 
-        writer.write(out);
+        Future<?> future = sstable.runWithReadLock(ignored -> {
+            CassandraEntireSSTableStreamWriter writer = new CassandraEntireSSTableStreamWriter(sstable, session, manifest);
+
+            // start concurrent sstable component mutation
+            Future<?> mutation = CompletableFuture.completedFuture(null);
+            if (componentMutation != null)
+            {
+                mutation = ScheduledExecutors.optionalTasks.submit(componentMutation);
+                // wait 1s for concurrent task to start
+                FBUtilities.sleepQuietly(1000);
```

Comment:
nit: Rather than a timed wait, what about just counting down on a latch when the provided `Callable` actually starts. (We'd probably have to make it a field in the test class, but it can be reset in a `@Before` block, etc.)

Reply:
+1